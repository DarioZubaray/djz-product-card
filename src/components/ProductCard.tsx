import React, { createContext, CSSProperties } from 'react';

import { useProduct } from '../hooks/useProduct';
import { ProductContextProps, Product, OnChangeArgs, InitialValues, ProductCardHandlers } from '../interfaces/interfaces';

import styles from '../styles/styles.module.css';

export interface ProductCardProps {
    product: Product;
    // children: ReactElement | ReactElement [];
    children: ( args: ProductCardHandlers) => JSX.Element
    className?: string;
    style?: CSSProperties;
    onChange?: (args: OnChangeArgs) => void;
    value?: number;
    initialValues?: InitialValues;
}

export const ProductContext = createContext({} as ProductContextProps)
const { Provider } = ProductContext

export const ProductCard = ({ children, product, className, style, onChange, value, initialValues }: ProductCardProps) => {

    const { 
        counter,
        increaseBy,
        maxCount,
        isMaxCountReached,
        reset
    } = useProduct({ onChange, product, value, initialValues });

    return (
        <Provider value={{
            counter,
            increaseBy,
            product,
            maxCount
        }}>
            <div className={` ${ styles.productCard } ${ className } `} style={ style }>
                { 
                    children({
                        count: counter,
                        increaseBy,
                        isMaxCountReached,
                        maxCount: initialValues?.maxCount,
                        product,
                        reset
                    })
                }
            </div>
        </Provider>
    );
};

