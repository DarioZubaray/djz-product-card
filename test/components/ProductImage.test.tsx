import * as React from 'react';
import renderer from 'react-test-renderer';
import { ProductImage, ProductCard } from '../../src/components/';
import { product2 } from '../data/product';

describe('Product image', () => {
    test('debe de mostrar el componente con la imagen por personalizada', () => {

        const wrapper = renderer.create(
            <ProductImage img='algun-img-path'/>
        )
        console.log(wrapper.toJSON());
        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    test('debe de mostrar el componente con la imagen del producto', () => {
        const wrapper = renderer.create(
            <ProductCard product={ product2 } >
                {
                    () => (
                        <ProductImage />
                    )
                }
            </ProductCard>
        )

        expect(wrapper.toJSON()).toMatchSnapshot();
    })
});
