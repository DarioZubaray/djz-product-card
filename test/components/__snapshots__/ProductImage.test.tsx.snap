// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product image debe de mostrar el componente con la imagen del producto 1`] = `
<div
  className=" productCard undefined "
>
  <img
    alt="Product"
    className="productImg undefined"
    src="./coffee-mug2.png"
  />
</div>
`;

exports[`Product image debe de mostrar el componente con la imagen por personalizada 1`] = `
<img
  alt="Product"
  className="productImg undefined"
  src="algun-img-path"
/>
`;
